File: receivers/common/web/ListenerService.ts
import { FCastSession } from 'common/FCastSession';
import { Opcode, EventSubscribeObject, EventObject, EventType, KeyEvent, KeyDownEvent, KeyUpEvent } from 'common/Packets';
import { Logger, LoggerType } from 'common/Logger';
import { deepEqual } from 'common/UtilityBackend';
import { EventEmitter } from 'events';
import { errorHandler } from 'src/Main';
const logger = new Logger('ListenerService', LoggerType.BACKEND);

export abstract class ListenerService {
    public emitter: EventEmitter = new EventEmitter();
    protected sessionMap: Map<string, FCastSession> = new Map();
    private eventSubscribers: Map<string, EventSubscribeObject[]> = new Map();

    public abstract start(): void;
    public abstract stop(): void;
    public abstract disconnect(sessionId: string): void;

    public send(opcode: number, message = null, sessionId = null) {
        // logger.info(`Sending message ${JSON.stringify(message)}`);

        if (sessionId) {
            if (opcode === Opcode.Event.valueOf() && !this.isSubscribedToEvent(sessionId, message.event)) {
                return;
            }

            try {
                this.sessionMap.get(sessionId)?.send(opcode, message);
            } catch (e) {
                logger.warn("Failed to send error.", e);
                this.sessionMap.get(sessionId)?.close();
            }
        }
        else {
            for (const session of this.sessionMap.values()) {
                if (opcode === Opcode.Event.valueOf() && !this.isSubscribedToEvent(session.sessionId, message.event)) {
                    continue;
                }

                try {
                    session.send(opcode, message);
                } catch (e) {
                    logger.warn("Failed to send error.", e);
                    session.close();
                }
            }
        }
    }

    public subscribeEvent(sessionId: string, event: EventSubscribeObject) {
        if (!this.eventSubscribers.has(sessionId)) {
            this.eventSubscribers.set(sessionId, []);
        }

        let sessionSubscriptions = this.eventSubscribers.get(sessionId);
        sessionSubscriptions.push(event);
        this.eventSubscribers.set(sessionId, sessionSubscriptions);
    }

    public unsubscribeEvent(sessionId: string, event: EventSubscribeObject) {
        if (this.eventSubscribers.has(sessionId)) {
            let sessionSubscriptions = this.eventSubscribers.get(sessionId);

            const index = sessionSubscriptions.findIndex((obj) => deepEqual(obj, event));
            if (index != -1) {
                sessionSubscriptions.splice(index, 1);
            }

            this.eventSubscribers.set(sessionId, sessionSubscriptions);
        }
    }

    public getSessions(): string[] {
        return [...this.sessionMap.keys()];
    }

    public getSessionProtocolVersion(sessionId: string) {
        return this.sessionMap.get(sessionId)?.protocolVersion;
    }

    public getAllSubscribedKeys(): { keyDown: Set<string>, keyUp: Set<string> } {
        let keyDown = new Set<string>();
        let keyUp = new Set<string>();

        for (const session of this.eventSubscribers.values()) {
            for (const event of session) {
                switch (event.type) {
                    case EventType.KeyDown:
                        keyDown = new Set([...keyDown, ...(event as KeyDownEvent).keys]);
                        break;

                    case EventType.KeyUp:
                        keyUp = new Set([...keyUp, ...(event as KeyUpEvent).keys]);
                        break;

                    default:
                        break;
                }
            }
        }

        return { keyDown: keyDown, keyUp: keyUp };
    }

    private isSubscribedToEvent(sessionId: string, event: EventObject): boolean {
        let isSubscribed = false;

        if (this.eventSubscribers.has(sessionId)) {
            for (const e of this.eventSubscribers.get(sessionId).values()) {
                if (e.type === event.type) {
                    if (e.type === EventType.KeyDown.valueOf() || e.type === EventType.KeyUp.valueOf()) {
                        const subscribeEvent = e.type === EventType.KeyDown.valueOf() ? e as KeyDownEvent : e as KeyUpEvent;
                        const keyEvent = event as KeyEvent;

                        if (!subscribeEvent.keys.includes(keyEvent.key)) {
                            continue;
                        }
                    }

                    isSubscribed = true;
                    break;
                }
            }
        }

        return isSubscribed;
    }

    protected async handleServerError(err: NodeJS.ErrnoException) {
        errorHandler(err);
    }
}

File: receivers/android/app/src/main/java/com/futo/fcast/receiver/ListenerService.kt
package com.futo.fcast.receiver

import android.util.Log
import com.futo.fcast.receiver.models.EventMessage
import com.futo.fcast.receiver.models.EventObject
import com.futo.fcast.receiver.models.EventSubscribeObject
import com.futo.fcast.receiver.models.EventType
import com.futo.fcast.receiver.models.KeyDownEvent
import com.futo.fcast.receiver.models.KeyEvent
import com.futo.fcast.receiver.models.KeyUpEvent
import com.futo.fcast.receiver.models.Opcode
import java.util.UUID

abstract class ListenerService {
    protected val sessionMap: MutableMap<UUID, FCastSession> = mutableMapOf()
    private val _eventSubscribers: MutableMap<UUID, MutableList<EventSubscribeObject>> = mutableMapOf()

    abstract fun start()
    abstract fun stop()
    abstract fun disconnect(sessionId: UUID)

    fun <T> send(opcode: Opcode, message: T? = null, sessionId: UUID? = null) {
//        Log.i(TAG, "Sending message $message")

        if (sessionId != null) {
            if (opcode == Opcode.Event && !this.isSubscribedToEvent(sessionId, (message as EventMessage).event)) {
                return
            }

            try {
                this.sessionMap[sessionId]?.send(opcode, message)
            } catch (e: Throwable) {
                Log.w(TAG, "Failed to send error.", e)
                this.sessionMap[sessionId]?.close()
            }
        }
        else {
            for (session in this.sessionMap.values) {
                if (opcode == Opcode.Event && !this.isSubscribedToEvent(session.id, (message as EventMessage).event)) {
                    continue
                }

                try {
                    session.send(opcode, message)
                } catch (e: Throwable) {
                    Log.w(TAG, "Failed to send error.", e)
                    session.close()
                }
            }
        }
    }

    fun subscribeEvent(sessionId: UUID, event: EventSubscribeObject) {
        val sessionSubscriptions = _eventSubscribers.getOrDefault(sessionId, mutableListOf<EventSubscribeObject>())
        sessionSubscriptions += event
        _eventSubscribers[sessionId] = sessionSubscriptions
    }

    fun unsubscribeEvent(sessionId: UUID, event: EventSubscribeObject) {
        if (_eventSubscribers.containsKey(sessionId)) {
            val sessionSubscriptions = _eventSubscribers[sessionId]
            sessionSubscriptions?.remove(event)
            _eventSubscribers[sessionId] = sessionSubscriptions!!
        }
    }

    fun getSessions(): List<UUID> {
        return sessionMap.keys.toList()
    }

    fun getSessionProtocolVersion(sessionId: UUID): Long? {
        return sessionMap[sessionId]?.protocolVersion
    }

    fun getAllSubscribedKeys(): Pair<Set<String>, Set<String>> {
        val keyDown = mutableSetOf<String>()
        val keyUp = mutableSetOf<String>()

        for (session in _eventSubscribers.values) {
            for (event in session) {
                when (event.type) {
                    EventType.KeyDown -> keyDown += (event as KeyDownEvent).keys
                    EventType.KeyUp -> keyUp += (event as KeyUpEvent).keys
                    else -> {}
                }
            }
        }

        return Pair(keyDown, keyUp)
    }

    private fun isSubscribedToEvent(sessionId: UUID, event: EventObject): Boolean {
        var isSubscribed = false

        if (_eventSubscribers.containsKey(sessionId)) {
            for (e in _eventSubscribers[sessionId] ?: arrayListOf()) {
                if (e.type == event.type) {
                    when (e.type) {
                        EventType.KeyDown -> {
                            val subscribeEvent = e as KeyDownEvent
                            val keyEvent = event as KeyEvent
                            if (!subscribeEvent.keys.contains(keyEvent.key)) {
                                continue
                            }
                        }
                        EventType.KeyUp -> {
                            val subscribeEvent = e as KeyUpEvent
                            val keyEvent = event as KeyEvent
                            if (!subscribeEvent.keys.contains(keyEvent.key)) {
                                continue
                            }
                        }
                        else -> {}
                    }

                    isSubscribed = true
                    break
                }
            }
        }

        return isSubscribed
    }
// TODO: Address error handling for UI 2.0
//    protected async handleServerError(err: NodeJS.ErrnoException) {
//        errorHandler(err)
//    }

    companion object {
        private const val TAG = "ListenerService"
    }
}